### DISCUSSION TITLE
Litematic export as old Schematic

### DISCUSSION THREAD

**Discussion by sleepytyler** [maruohon/litematica#332] [Q&A] [answered]
Are we ever going to see a export as schematic ability for this mod? 

Is there any way currently to convert either the NBT or litematic files to baritone compatible .schematics?

**Comment by maruohon** [comment, marked as answer]
Yes the intention is for Litematica to support all the common schematic formats. The 1.12.2 version has had support for all 4 formats for over a year already. Those formats being its own `*.litematic`, the Sponge schematic format `*.schem` (used by World Edit in 1.13+, and possibly Baritone?), the vanilla structure block template `*.nbt` and the old MCEdit/Schematica format `*.schematic`.

So currently the 1.13.2+ versions of Litematica are unfortunately still based on the main development code from June 2019, so the newer improved schematic handling code is not yet included in those versions. And also the vanilla "flattening" that happened in 1.13 creates additional problems. Even converting schematics in the "modern formats" (ie. anything except the old MCEdit format) needs special conversion code to support converting schematics across the 1.12.2 <-> 1.13+ MC version boundary. And the flattening is especially problematic for trying to support the old MCEdit/Schematica format, and in fact none of the new 1.13+ blocks can be saved in that old format in a "generic and reliable way", because any blocks that have any block state properties can not be represented in that format, even with a **block** ID palette (it would need a block state palette instead of just a block ID).

So basically support is planned and also partially implemented for converting between any of these formats (as far as possible). But most of that code is currently only in 1.12.2, and the cross-MC version conversion code is not yet ready and isn't even included in the 1.12.2 releases yet.

I don't now which format Baritone actually supports. The new Sponge format is unfortunately often also named as `*.schematic` instead of `*.schem`, which creates confusion with the old totally different MCEdit/Schematica format. But none of the new 1.13+ blocks can be saved in the MCEdit schematic format, so I assume it must support the Sponge format?

But basically at least full support for the Sponge `*.schem` format is coming, and also as good as possible support for the old MCEdit `*.schematic` format is coming, although it's not recommended to save anything to that format from 1.13+. I'm not sure about the timeline, there is still a lot of work left before the next releases are ready.

**Reply by sleepytyler** [reply]
I appreciate you taking the time out of your day to address this, I know that its not as easy as mapping x->y in a few dozen lines of code so I know these things will take time and effort. I was not however aware that there was a 1.12.2 version of litematica, curseforge does not seem to have a reference to it [https://www.curseforge.com/minecraft/mc-mods/litematica/files/all](url) and there are no releases listed on this github project so where could I get a copy of an official release for this (fabric if possible)? 

Also if you have a discord could you post it here? Ive been looking for that as well to no avail.

As for baritone I believe it currently supports .schem and .schematica (sponge/schematica respectively) and there is currently a merge request to implement support for .litematica as well seen here [https://github.com/cabaletta/baritone/pull/2544/commits/5d7128ba55ef072d5a54e917f467334a1a76ee25](url). 

Shortly after posting the previous question I found the litematica-printer mod that seems to help bypass the need to use baritone for anyone who might be reading this in the future, however I still have yet to find a reliable way to convert the older schematics to newer ones.

**Reply by maruohon** [reply]
The 1.12 versions are currently still only on my own server, here: https://masa.dy.fi/mcmods/client_mods/
I will probably start uploading them to CurseForge after the next round of bigger updates of all the client mods are ready for release. The very latest builds of the mods are also there, and I usually only upload to CurseForge after a larger number of new features have been added (ie. not after each and every single feature new build), or if there was a bigger update or some important bug/crash fix.

Currently there are no 1.12.2 Legacy-Fabric versions yet, but those will likely come after the next big updates are ready at some point. The main development version and thus the version with the most features is the 1.12.2 LiteLoader version. The 1.12.2 Forge ports are only made occasionally, and thus far only twice.

On my channel on the SciCraft Discord there is also a pinned message with a temporary WIP build of Litematica, which has preliminary support for converting schematics between MC versions. So that version could probably be used for converting a 1.16 `*.litematic` into a 1.16 `*.schem` (Sponge schematic), although I don't think I have tested that yet. I've mostly used that build thus far to backport some `*.litematic` schematics from 1.13+ to 1.12.2 in the `*.litematic` format.

Note that the conversions currently still mostly only work between/using the Litematica or Sponge formats, as those are the most similar and easiest formats to work with. Especially the old MCEdit format `*.schematic` is going to be the most problematic when converting to or from 1.13+ schematics. Or well, there is actually support for loading old MCEdit schematics already in the 1.13+ versions of Litematica, so going forward already works there. And actually I don't think it works in that special build yet at all(?), since it does not use the same conversion code at all. And I don't remember if the vanilla format is supported at all or how well in that build. Like I said that conversion code is still very much WIP.

As for Discord and feedback, I have my own Discord server at https://discord.gg/2FgywHj, but the `#masas-mods` channel on the SciCraft Discord (https://discord.gg/scicraft) is usually the most common place for help/questions/feedback related to my client mods.

**Comment by snoweuph** [comment]
Is there support for converting to .schem in 1.16.6 yet?

**Reply by maruohon** [reply]
No, export to the Sponge schematic format i.e. `*.schem` is still only in the 1.12.2 version. There hasn't been that much progress to Litematica's main development version in 1.12.2  in over 2 years now, other than porting it to work with the refactored malilib. And thus it's still not ready to be "properly" ported to the other MC versions yet.

**Reply by snoweuph** [reply]
welp, actaully baritone just added full litematica support, its not in the releases yet, but in the CI builds
